ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader_builder.cpp

```diff
@@ -0,0 +1,124 @@
+#include "uploader_builder.hpp"
+#include "libdatadog_helpers.hpp"
+#include "types.hpp"
+
+#include <algorithm>
+#include <mutex>
+#include <string_view>
+
+using namespace Datadog;
+
+UploaderBuilder&
+UploaderBuilder::set_env(std::string_view _env)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_env.empty())
+        env = _env;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_service(std::string_view _service)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_service.empty())
+        service = _service;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_version(std::string_view _version)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_version.empty())
+        version = _version;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_runtime(std::string_view _runtime)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_runtime.empty())
+        this->runtime = _runtime;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_runtime_version(std::string_view _runtime_version)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_runtime_version.empty())
+        runtime_version = _runtime_version;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_profiler_version(std::string_view _profiler_version)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_profiler_version.empty())
+        profiler_version = _profiler_version;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_url(std::string_view _url)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_url.empty())
+        url = _url;
+    return *this;
+}
+UploaderBuilder&
+UploaderBuilder::set_tag(std::string_view _key, std::string_view _val)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_key.empty() && !_val.empty())
+        user_tags[_key] = _val;
+    return *this;
+}
+
+UploaderBuilder&
+UploaderBuilder::set_runtime_id(std::string_view _runtime_id)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    if (!_runtime_id.empty())
+        runtime_id = _runtime_id;
+    return *this;
+}
+
+std::unique_ptr<Uploader>
+UploaderBuilder::build_ptr()
+{
+    // Setup the ddog_Exporter
+    ddog_Vec_Tag tags = ddog_Vec_Tag_new();
+
+    // Add the tags, emitting the first failure
```

Comment:
where do we emit the failure?

Reply:
I refactored this.  Now we check _all_ the tags and send a big blob of all the failures, with the understanding that the average user will only have to fix one thing, but the _really annoyed_ user will have to fix several things iteratively.

How we log this is still a problem though.